runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,365 @@
+---@param ctx lsp.HandlerContext
+local function handler(err, result, ctx)
+  if err then
+    log.error(err)
+    return
+  end
+
+  if not result then
+    return
+  end
+
+  local bufnr = assert(ctx.bufnr)
+  local bufstate = assert(bufstates[bufnr])
+
+  ---@type table<integer, [integer, ">" | "<"?]?>
+  local row_level = {}
+  ---@type table<integer, string>
+  local row_text = {}
+  ---@type table<lsp.FoldingRangeKind, lsp.FoldingRange[]?>>
+  local kind_ranges = {}
+
+  for _, range in ipairs(result) do
+    local kind = range.kind
+    if kind then
+      ---@cast kind lsp.FoldingRangeKind
+      local ranges = kind_ranges[kind] or {}
+      ranges[#ranges + 1] = range
+      kind_ranges[kind] = ranges
+    end
+
+    local start_row = range.startLine
+    local end_row = range.endLine
+    -- Fold a single line is not supported.
+    if start_row ~= end_row then
+      row_text[start_row] = range.collapsedText
+      for row = start_row, end_row do
+        local level = row_level[row] or { 0 }
+        level[1] = level[1] + 1
+        row_level[row] = level
+      end
+      row_level[start_row][2] = '>'
+      row_level[end_row][2] = '<'
+    end
+  end
+
+  bufstate.row_level = row_level
+  bufstate.row_text = row_text
+  bufstate.kind_ranges = kind_ranges
+  bufstate.version = ctx.version
+
+  if api.nvim_get_mode().mode:match('^i') then
+    -- `foldUpdate()` is guarded in insert mode.
+    schedule_foldupdate(bufnr)
+  else
+    foldupdate(bufnr)
+  end
+end
+
+---@param bufnr integer
+---@param client vim.lsp.Client The client whose server supports `foldingRange`.
+---@return integer? request_id
+local function refresh(bufnr, client)
+  ---@type lsp.FoldingRangeParams
+  local params = { textDocument = util.make_text_document_params(bufnr) }
+  local _, request_id = client:request(ms.textDocument_foldingRange, params, handler, bufnr)
+  return request_id
+end
+
+-- MAINTENANCE NOTE:
+-- * `bufstate` requires event hooks to ensure correctness.
+-- * event hooks requires the previous `bufstate` for updates.
+--
+-- In other words, they are interdependent.
+-- Since their creation and destruction are handled manually,
+-- our strategy is to ensure that their lifecycles are always synchronized.
+--
+-- As they are only meaningful for buffers associated with
+-- at least one window that have `foldexpr()` set,
+-- the following three elements are guaranteed
+-- to exist or not exist simultaneously:
+--
+-- * A buffer associated with at least one window where `foldexpr()` is set.
+-- * Its `bufstate`.
+-- * Its event hooks.
+local augroup_setup = api.nvim_create_augroup('vim_lsp_folding_range/setup', {})
+
+--- Setup `bufstate` and its event hooks.
+---@param bufnr integer
+---@return vim.lsp.folding_range.BufState
+local function setup(bufnr)
+  ---@type vim.lsp.folding_range.BufState
+  local bufstate = { row_level = {}, row_text = {}, kind_ranges = {} }
+  if not api.nvim_buf_is_loaded(bufnr) then
+    return bufstate
+  end
```

Comment:
Can't you check for available clients here and return if there are none, this avoids setting up attaching to the buffer and setting up autocmds, and I think allows you to remove the OptionSet autocmd.

Reply:
If garbage collection is needed, `OptionSet` is necessary. Because the user may have changed `foldexpr` themselves, there is no way to know that the user has changed it other than using `OptionSet`. At the same time, LSP clients will not make any changes in this case.